pkg/scheduler/framework/interface.go

```diff
@@ -588,7 +588,7 @@ type Framework interface {
 	// cycle is aborted.
 	// It also returns a PreFilterResult, which may influence what or how many nodes to
 	// evaluate downstream.
-	RunPreFilterPlugins(ctx context.Context, state *CycleState, pod *v1.Pod) (*PreFilterResult, *Status)
+	RunPreFilterPlugins(ctx context.Context, state *CycleState, pod *v1.Pod, diagnosis *Diagnosis) (*PreFilterResult, *Status)
```

Comment:
another option would be to add UnschedulablePlugins to the PreFilterResult and have the caller of RunPrefilterPlugins use that result to update diagnosis. It avoids changing the interface, and makes the data flow clearer.

Reply:
Yes, we could do that.
Me myself prefer the current implementation.

Reply:
Yes, we can do that.
Me myself prefer the current implementation.

Reply:
Is there an advantage to the current implementation that I'm not seeing which justifies having side effects? My concern is that readers/maintainers of this code will have to read the implementation of `RunPreFilterPlugins` to see why `Diagnosis` is passed, and which fields in it are read/mutated. Passing this type also makes it easier in the future for other fields in `Diagnosis` to be modified without careful consideration, increasing complexity (suppose `NodeToStatusMap` is also modified inside of `RunPreFilterPlugins`, rather than just by [findNodesThatFitPod](https://github.com/kubernetes/kubernetes/blob/b0263baae51895058b64b8e6f510a22a14ed8d3d/pkg/scheduler/schedule_one.go#L453)).

While the data flow/mutability is my primary concern, this PR also becomes smaller (8 -> 4 files), if we include result in `PreFilterResult`

Reply:
I agree with @gabesaba's rationale.
Please update the PR.

Reply:
already updated, please review again @AxeZhan @gabesaba @alculquicondor

Reply:
Can we just return UnschedulablePlugins from `RunPreFilterPlugins`, instead of a new field in `PreFilterResult `?

`PreFilterResult` is a data structure we use between the framework and plugins, which is supposed to be exposed to plugin developers, while `UnschedulablePlugins` is not a field for plugins developers.

Reply:
@gabesaba has raised a similar point. https://github.com/kubernetes/kubernetes/pull/122251#discussion_r1637720634

We can do this as a follow-up?

Reply:
I believe this should be done in this PR.

Reply:
@alculquicondor @AxeZhan @gabesaba Do as @sanposhiho  say?

Reply:
If you have time doing this, then I'm +1 for it.
Apologize for the frequent changes.

Reply:
done as  @sanposhiho say